### PR TITLE
[build] Move all Find*.cmake to cmake/Modules

### PR DESCRIPTION
**Pull request by orionr**
* Move remaining `Find*.cmake` files into `cmake/Modules`
* Note that `FindGloo.cmake` and `FindNCCL.cmake` were different, but I'm tempted to take the PyTorch versions, which is what this does. We'll see how Caffe2 builds fare.

c @ezyang, @Yangqing, @pjh5 for review. Thanks.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
It'd be good to check the history of the Find files to see if there are any relevant looking updates.

**Comment by orionr** [comment]
Good point - I'll look at the file histories and see if there was anything interesting added for either of them.

File: cmake/Modules/FindGloo.cmake [hunk 1]
```diff
@@ -1,45 +1,22 @@
-# Try to find the Gloo library and headers.
-#  Gloo_FOUND        - system has Gloo lib
-#  Gloo_INCLUDE_DIRS - the Gloo include directory
-#  Gloo_LIBRARIES    - libraries needed to use Gloo
+IF(NOT GLOO_FOUND)
```
Comment (orionr) [inline cmake/Modules/FindGloo.cmake]: Looks like the two FindGloo.cmake histories are

Caffe2 - single commit May 2017
https://github.com/pytorch/pytorch/commits/a2a28c0ef1d9a433972fe72fa5b0b9b850ccfcaf/cmake/Modules/FindGloo.cmake

PyTorch - a couple of commits March - May 2017
https://github.com/pytorch/pytorch/commits/4ebf3ff46dc0d4899fb163a970ba4dc68c566b2a/torch/lib/THD/cmake/FindGloo.cmake

@apaszke and @pietern seemed to have touched these. Which do you think should win and/or how should we merge?

File: cmake/Modules/FindNCCL.cmake [hunk 2]
```diff
@@ -1,28 +1,20 @@
-# Find the nccl libraries
+# - Try to find NCCL
```
Comment (orionr) [inline cmake/Modules/FindNCCL.cmake]: And the two FindNCCL.cmake histories are

Caffe2 - multiple commits Aug 2016 - April 2018
https://github.com/pytorch/pytorch/commits/108f5c197fa71883948ea95ec0d5a609e737c066/cmake/Modules/FindNCCL.cmake

PyTorch - two commits Nov 2017 - Apr 2018
https://github.com/pytorch/pytorch/commits/108f5c197fa71883948ea95ec0d5a609e737c066/torch/lib/THD/cmake/FindNCCL.cmake

Looks like @pietern , and @soumith touched these. Also any thoughts on which should be used going forward?

**Comment by pietern** [comment]
@orionr  Thanks for looking up the histories. I think we're fine to stick with the PyTorch versions, but do take a look at the find_library calls in FindNCCL for Caffe2 and ensure the directories listed there are ported over to the PyTorch one (see https://github.com/pytorch/pytorch/commit/54342287fed769021543c0197a836eec94dd351d#diff-5df661110abfc4edc2e80c967077e235 for explanation)

File: cmake/Modules/FindGloo.cmake (hunk 1 above)
Comment (orionr) [reply cmake/Modules/FindGloo.cmake]: In this case I'm going to take the Caffe2 version. @apaszke let me know if you think otherwise.

Comment (apaszke) [reply cmake/Modules/FindGloo.cmake]: The Caffe2 looks nicer to me, so I think we should go with that one.

File: cmake/Modules/FindNCCL.cmake [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Find the nccl libraries
```
Comment (pjh5) [inline cmake/Modules/FindNCCL.cmake]: For all these changes you'll have to recreate the docker images first

Comment (pietern) [reply cmake/Modules/FindNCCL.cmake]: Why exactly?

Comment (pjh5) [reply cmake/Modules/FindNCCL.cmake]: Actually after looking further I don't think that this is the problem.

**Comment by ezyang** [comment]
Failure looks legit-ish?

```
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/GlooCache.hpp: In instantiation of 'static thd::GlooCache::key_type thd::gloo_cache::algorithm_spec<(thd::CollectiveType)3u, T>::key(THDGroup, thd::DeviceType, std::size_t, std::size_t, THDReduceOp) [with T = float; thd::GlooCache::key_type = std::tuple<thd::CollectiveType, int, thd::DeviceType, int, long unsigned int, long unsigned int, THDReduceOp, unsigned int>; THDGroup = int; std::size_t = long unsigned int]':
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/GlooCache.hpp:158:53:   required from 'thd::GlooCache::value_type thd::GlooCache::getAlgorithm(THDGroup, const thd::DataChannelGloo::Group&, Args ...) [with thd::CollectiveType D = (thd::CollectiveType)3u; T = float; Args = {thd::DeviceType, long unsigned int, long int, THDReduceOp}; thd::GlooCache::value_type = std::tuple<std::shared_ptr<gloo::Algorithm>, std::shared_ptr<char>, std::shared_ptr<char>, std::shared_ptr<std::mutex> >; THDGroup = int]'
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/DataChannelGloo.cpp:223:89:   required from 'void thd::DataChannelGloo::allReduceT(at::Tensor&, THDReduceOp, THDGroup) [with T = float; THDGroup = int]'
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/DataChannelGloo.cpp:236:3:   required from here
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/GlooCache.hpp:300:67: error: 'THDGetCudaState' was not declared in this scope
19:06:38        auto cuda_stream = THCState_getCurrentStream(THDGetCudaState());
19:06:38                                                                    ^
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/GlooCache.hpp:300:51: error: 'THCState_getCurrentStream' was not declared in this scope
19:06:38        auto cuda_stream = THCState_getCurrentStream(THDGetCudaState());
19:06:38                                                    ^
19:06:38 /var/lib/jenkins/workspace/torch/lib/THD/base/data_channels/GlooCache.hpp:301:30: error: 'THDGetStreamId' was not declared in this scope
19:06:38        stream = THDGetStreamId(cuda_stream);
```

Also needs rebase.

**Pull request closed without merging**